mass_mailing_partner/models/mail_mass_mailing_contact.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2015 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# Copyright 2015 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# Copyright 2015 Javier Iniesta <javieria@antiun.com>
+# Copyright 2017 David Vidal <david.vidal@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class MailMassMailingContact(models.Model):
+    _inherit = 'mail.mass_mailing.contact'
+
+    partner_id = fields.Many2one(comodel_name='res.partner', string="Partner",
+                                 domain=[('email', '!=', False)])
+
+    @api.constrains('partner_id', 'list_ids')
+    def _check_partner_id_lists_ids(self):
+        for contact in self:
+            if contact.partner_id:
+                other_contact = self.search([
+                    ('partner_id', '=', contact.partner_id.id),
+                    ('id', '!=', contact.id)
+                ])
+                if contact.list_ids & other_contact.mapped('list_ids'):
+                    raise ValidationError(_("Partner already exists in one of "
+                                            "these mailing lists"))
+
+    @api.onchange('partner_id')
+    def _onchange_partner(self):
+        if self.partner_id:
+            self.name = self.partner_id.name
+            self.email = self.partner_id.email
+            if self.partner_id.title:
+                self.title_id = self.partner_id.title.id
+            if self.partner_id.company_id:
+                self.company_name = self.partner_id.company_id.name
+            if self.partner_id.country_id:
+                self.country_id = self.partner_id.country_id.id
+            if self.partner_id.category_id:
+                self.tag_ids = [(6, 0, self.partner_id.category_id.ids)]
+
+    @api.model
+    def create(self, vals):
+        if not vals.get('partner_id'):
+            vals = self._set_partner(vals)
+        vals = self._set_name_email(vals)
+        return super(MailMassMailingContact, self).create(vals)
+
+    def write(self, vals):
+        for contact in self:
+            if vals.get('partner_id', None) is False:
+                # If removing partner, search again by email
+                vals = contact._set_partner(vals)
+            vals = contact._set_name_email(vals)
+        return super(MailMassMailingContact, self).write(vals)
+
+    def _prepare_partner(self, vals, list_ids):
+        partner_vals = {
+            'name': vals.get('name', self.name) or vals.get('email',
+                                                            self.name),
+            'email': vals.get('email', self.email),
+            'country_id': vals.get('country_id',
+                                   self.country_id and self.country_id.id),
```

Comment:
`self.country_id.id` will be `False` if `len(self.country_id) == 0`, so no need to to this kind of trick.